.github/workflows/e2e.yaml

```diff
@@ -17,17 +17,26 @@ jobs:
           - v1.21.2
       max-parallel: 2
     runs-on: ubuntu-latest
+    permissions:
+      packages: write  # To be able to push images.
+
     steps:
       - name: Setup Go
         uses: actions/setup-go@v2
         with:
-          go-version: ^1.17
+          go-version: 1.17.x
 
       - name: Check out code into the Go module directory
         uses: actions/checkout@v2
         with:
           submodules: recursive
 
+      - name: Login to container registry
```

Comment:
Please use the [container registry](https://github.com/shipwright-io/cli/blob/main/test/e2e/helpers.sh) inside the [KinD instance](https://github.com/shipwright-io/cli/blob/d26bd20bc2fafc3dc17dd1f73f3f251ecb370579/Makefile#L18-L23) as well.

Reply:
> > Please use the [container registry](https://github.com/shipwright-io/cli/blob/main/test/e2e/helpers.sh) inside the [KinD instance](https://github.com/shipwright-io/cli/blob/d26bd20bc2fafc3dc17dd1f73f3f251ecb370579/Makefile#L18-L23) as well.
> 
> Thanks for bringing this up, I was meaning to talk to you about that. Forgot to mention that in the PR text.
> 
> My problem is the following:

Let me split the issue on topics, so we can discuss the details.

> - The bundle approach is uploading the local source code as a container image to the registry from the location of the client.

Sure!

> - This means, by default, the container registry hosted by the cluster is not reachable unless we open some port to make it available.

Currently, the [port 32222 is open](https://github.com/shipwright-io/cli/blob/d26bd20bc2fafc3dc17dd1f73f3f251ecb370579/hack/install-registry.sh#L61-L66) during testing, please consider.

> - Even then, the problem is the name of the image. I would need to push the image to a localhost-named registry, but it is suppose to be consumed with the other name in the source step. In the Shipwright build E2E tests we have some hacks in the test suite itself to deal with that.

That can be easily achieved by manipulating the `/etc/hosts` file, alternatively using [localtest.me](https://readme.localtest.me/) type of service.

> - However, I did not see a "good" way to do it in the CLI, too. I would have to have a "hack" of sorts in the CLI code itself, code that is only suppose to be used for test cases to to speak and I do not like to be honest.

Yeah, lets work on this testing scenario to make sure we don't need such thing.

> - The other option would have been to fiddle with hostname alias or DNS-hacks to be able to push to a registry with a cluster internal hostname from the client.

Yes, that's precisely my suggestion here.

> - And at this point I figured we use the GitHub container registry in Shipwright Build already, it is available to us in the GitHub Actions, you can use it locally when testing, so why not try to use it here as well.

All end-to-end tests in this repository depend on a container registry, which you can set as a [Makefile target parameter as well](https://github.com/shipwright-io/cli/blob/d26bd20bc2fafc3dc17dd1f73f3f251ecb370579/Makefile#L21-L23). So, lets make sure the newly added functionality uses the same testing infrastructure.

Reply:
I used a combination of `/etc/hosts` and kubect port forwarding. It looks like the local node port does not work in a GitHub Action setup as the network type is different. On my local machine, it worked perfectly fine with just the hostname entry, but in GitHub Actions it failed. So I tried different options and eventually decided to just do a port forwarding.

Reply:
We will need the `/etc/hosts` for the synthetic hostname resolution, good move! I've tweak the approach to expose the port 32222 on https://github.com/shipwright-io/cli/pull/106, please consider.